src/DurableTask.Netherite/OrchestrationService/NetheriteOrchestrationService.cs

```diff
@@ -347,20 +359,39 @@ public async Task TryStartAsync(bool clientOnly)
             }
         }
 
-        async Task<ServiceState> StartClientAsync(Task<bool> greenLight)
+        static void TryTransition<T>(ref Task<T> transition, Task<T> lastTransition, Func<Task<T>> nextTransition)
         {
-            if (!await greenLight) return ServiceState.None; 
+            var greenLight = new TaskCompletionSource<bool>();
+            var task = conditionalTransition();
+            var interlockedResult = Interlocked.CompareExchange<Task<T>>(ref transition, task, lastTransition);
+            bool success = interlockedResult == lastTransition;
+            greenLight.SetResult(success);
```

Comment:
I think we should provide guiding comments here, as this is the key change from what I gather.

```suggestion

            // construct a "green light" task to control whether or not we're allowed to take the next transitoin step
            var greenLight = new TaskCompletionSource<bool>();
            var task = conditionalTransition();
            
            // If the current transition task is the same as `lastTransition`, the transition at the beginning of this operation
            // then it is safe to invoke the next transition task. If so, we set `transition` to `task`, the next transition.
            var detectedTransition = Interlocked.CompareExchange<Task<T>>(ref transition, task, lastTransition);
            bool isSafeToTransition = detectedTransition  == lastTransition;
            greenLight.SetResult(isSafeToTransition);
```